Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py

```diff
@@ -311,14 +318,24 @@ def is_checkout(cls, path):
         return run([cls.executable(), 'rev-parse', '--show-toplevel'], cwd=path, capture_output=True).returncode == 0
 
     @decorators.hybridmethod
-    def config(context):
+    @decorators.Memoize()
+    def config(context, location=None):
         args = [context.executable(), 'config', '-l']
         kwargs = dict(capture_output=True, encoding='utf-8')
+        if location and location not in context.CONFIG_LOCATIONS:
+            raise TypeError("'{}' is not a valid git config location".format(location))
 
-        if isinstance(context, type):
+        if isinstance(context, type) or location == 'global':
+            if location in ['repository', 'project']:
+                raise TypeError("Cannot find '{}' git config without repository".format(location))
```

Comment:
I read this as if the location is a repository or project raise an error: cannot find the config without a repository.  I'm guessing that not what this code does so maybe some variables need renames?

Reply:
Also if location == global how would it match repository or project?

Reply:
Your initial reading is correct.

There are two ways this function can be called:
```
Git.config()
```
and
```
Git(<path>).config()
```
In the first case, you're calling `config` as a `classmethod`, so you don't have either a repository or project config. That's exactly what the `isinstance(context, type)` is checking for. If we're calling `config` as a `classmethod`, or we're calling it as an instance method but specifically asking for global state, then we do the check to make sure we aren't being asked for something we don't have as a `classmethod`

would this possibly be less confusing?
```
if isinstance(context, type) and location in ['repository', 'project']:
    raise TypeError("Cannot find '{}' git config without repository".format(location))
if isinstance(context, type) or location == 'global':
    args += ['--global']
else:
...
```

Reply:
The confusion was more that if the location is a repository then it doesn't make any sense to me that it would not have a repository.  So I guess the deeper question is how can we know the location is a repository if it doesn't have one?

Reply:
Your `git config` can com from 3 places:
`global` -> `~/.gitconfig`
`repository` -> `<checkout>/.git/config`
`project` -> `<checkout>/metadata/project_config` (note that this is added in this change)

For both the `repository` and `project` locations, we need to know where your checkout is. So the answer to
> how can we know the location is a repository if it doesn't have one?
is that if we get asked for `location='repository'` but don't actually provide something that defines where that repository is, we raise `TypeError("Cannot find 'repository' git config without repository")`, which as I type it out, should probably be `TypeError("Cannot find 'repository' git config without local checkout")`